File: Project_Go_Stack_by_ListNode/stackspeedtest/stack/stack.go
package stack

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

type Stack struct {
	Node   *ListNode
	Length int
}

func Constructor() *Stack {
	return &Stack{nil, 0}
}

func (this *Stack) Push(val int) {
	if this.Node == nil {
		this.Node = &ListNode{val, nil}
	} else {
		nextNode := *this.Node
		*this.Node = ListNode{val, &nextNode}
	}
	this.Length++
}

func (this *Stack) Pop() (int, error) {
	if this.Node == nil {
		return 0, fmt.Errorf("Stack is empty")
	}

	item := this.Node.Val
	if this.Node.Next != nil {
		*this.Node = *this.Node.Next
	} else {
		this.Node = nil
	}
	this.Length--

	return item, nil
}

func (this *Stack) Peek() (int, error) {
	if this.Node == nil {
		return 0, fmt.Errorf("Stack is empty")
	}

	return this.Node.Val, nil
}

func (this *Stack) Empty() bool {
	if this.Node == nil {
		return true
	}

	return false
}

func (this *Stack) Size() int {
	return (*this).Length
}

func (this *Stack) Search(val int) int {
	nodeTemp := this.Node
	for index := 1; nodeTemp != nil; index++ {
		if nodeTemp.Val == val {
			return index
		}
		nodeTemp = nodeTemp.Next
	}
	return -1
}

File: Project_C_Stack_by_ListNode/lib/stack.c
#include <time.h>
#include "../include/stack.h"

/**
    Implement Stack using Queues.
 */

/* Create a stack */
Stack* stackCreate(void) {
    Stack *stack = (Stack *)malloc(sizeof(Stack));

    if (stack == NULL) {
        return NULL;
    }

    /* I'm a stack implemented using linked queue, so I don't need maxSize */
    stack->size = 0;

    return stack;
}

/* Push element x onto stack */
void stackPush(Stack *stack, int element) {
    if (stack == NULL) {
        return;
    }

    ListNode *new_node = (ListNode *)malloc(sizeof(struct ListNode));
    if (new_node == NULL) {
        fprintf(stderr, "myStackPush(%d) new_node allocate error.", element);
        exit(-1);
    }

    new_node->val = element;
    new_node->next = stack->node;
    stack->node = new_node;

    stack->size++;
}

/* Removes the element on top of the stack */
int stackPop(Stack *stack) {
    if (stack == NULL) {
        return 0;
    }

    int item = stack->node->val;
    if (stack->node->next != NULL) {
        ListNode *tmp = stack->node;
        stack->node = stack->node->next;
        free(tmp);
    } else {
        free(stack->node);
        stack->node = NULL;
    }

    stack->size--;

    return item;
}

/* Get the top element */
int stackTop(Stack *stack) {
    if (stack == NULL) {
        return 0;
    }

    return stack->node->val;
}

/* Return whether the stack is empty */
bool stackEmpty(Stack *stack) {
    return (stack->size == 0) ? true: false;
}

/* Return val index */
int stackSearch(Stack *stack, int val) {
    ListNode *workNode = stack->node;

    for (int index = 0; workNode != NULL; index++) {
        if (workNode->val == val) {
            return index;
        }
        workNode = workNode->next;
    }

    return -1;
}

int stackSize(Stack *stack) {
    return stack->size;
}

/* Destroy the stack */
void stackDestroy(Stack *stack) {
    if (stack == NULL) {
        return;
    }

    if (stack->size > 0) {
        ListNode_free(stack->node);
    }

    free(stack);
}

int ListNode_free(struct ListNode* node) {
    if (node == NULL)
        return 0;
    
    int result = 0;

    if (node->next != NULL) {
        result += ListNode_free(node->next);
    }

    free(node);
    result++;

    return result;
}
